rust/operator-binary/src/service.rs

```diff
@@ -0,0 +1,117 @@
+pub fn build_rolegroup_headless_service(
+    trino: &v1alpha1::TrinoCluster,
+    role_group_ref: &RoleGroupRef<v1alpha1::TrinoCluster>,
+    object_labels: ObjectLabels<v1alpha1::TrinoCluster>,
+    selector: BTreeMap<String, String>,
+) -> Result<Service, Error> {
+    Ok(Service {
+        metadata: ObjectMetaBuilder::new()
+            .name_and_namespace(trino)
+            .name(rolegroup_headless_service_name(
+                &role_group_ref.object_name(),
+            ))
+            .ownerreference_from_resource(trino, None, Some(true))
+            .context(ObjectMissingMetadataForOwnerRefSnafu)?
+            .with_recommended_labels(object_labels)
+            .context(MetadataBuildSnafu)?
+            .build(),
+        spec: Some(ServiceSpec {
+            // Internal communication does not need to be exposed
+            type_: Some("ClusterIP".to_string()),
+            cluster_ip: Some("None".to_string()),
+            ports: Some(headless_service_ports(trino)),
+            selector: Some(selector),
+            publish_not_ready_addresses: Some(true),
+            ..ServiceSpec::default()
+        }),
+        status: None,
+    })
+}
+
+/// The rolegroup metrics [`Service`] is a service that exposes metrics and a prometheus scraping label.
+pub fn build_rolegroup_metrics_service(
+    trino: &v1alpha1::TrinoCluster,
+    role_group_ref: &RoleGroupRef<v1alpha1::TrinoCluster>,
+    object_labels: ObjectLabels<v1alpha1::TrinoCluster>,
+    selector: BTreeMap<String, String>,
+) -> Result<Service, Error> {
+    Ok(Service {
+        metadata: ObjectMetaBuilder::new()
+            .name_and_namespace(trino)
+            .name(rolegroup_metrics_service_name(
+                &role_group_ref.object_name(),
+            ))
+            .ownerreference_from_resource(trino, None, Some(true))
+            .context(ObjectMissingMetadataForOwnerRefSnafu)?
+            .with_recommended_labels(object_labels)
+            .context(MetadataBuildSnafu)?
+            .with_label(Label::try_from(("prometheus.io/scrape", "true")).context(LabelBuildSnafu)?)
+            .build(),
+        spec: Some(ServiceSpec {
+            // Internal communication does not need to be exposed
+            type_: Some("ClusterIP".to_string()),
+            cluster_ip: Some("None".to_string()),
+            ports: Some(metrics_service_ports()),
+            selector: Some(selector),
+            publish_not_ready_addresses: Some(true),
+            ..ServiceSpec::default()
+        }),
+        status: None,
+    })
+}
+
+fn headless_service_ports(trino: &v1alpha1::TrinoCluster) -> Vec<ServicePort> {
+    let name = trino.exposed_protocol().to_string();
+    let port = trino.exposed_port().into();
+
+    vec![ServicePort {
+        name: Some(name),
+        port,
+        protocol: Some("TCP".to_string()),
+        ..ServicePort::default()
+    }]
+}
```

Comment:
Is this for a particular Role? Eg: the Coordinator would likely expose ports for the workers to connect to, but the workers wouldn't expose the same ports (or do they too?).

I'm not sure how Trino clusters are formed, but the main thing is that we differentiate:
- listener service ports: things that are meant to be exposed to end users
- headless ports: anything for internal use (might be the same as listener ports)

Of course, maybe Trino is simple and it's all on a single port no matter which role, no matter if it's internal cluster forming, or for external usage.